packages/daimo-api/src/contract/keyRegistry.ts

```diff
@@ -0,0 +1,166 @@
+
+const signingKeyAddedEvent = getAbiItem({
+  abi: accountABI,
+  name: "SigningKeyAdded",
+});
+const signingKeyRemovedEvent = getAbiItem({
+  abi: accountABI,
+  name: "SigningKeyRemoved",
+});
+
+type SigningKeyAddedLog = Log<
+  bigint,
+  number,
+  false,
+  typeof signingKeyAddedEvent,
+  true
+>;
+type SigningKeyRemovedLog = Log<
+  bigint,
+  number,
+  false,
+  typeof signingKeyRemovedEvent,
+  true
+>;
+type SigningKeyAddedOrRemovedLog = SigningKeyAddedLog | SigningKeyRemovedLog;
+
+export class KeyRegistry {
+  /* In-memory indexer, for now. */
+  private keyToAddr = new Map<Hex, Address>();
+  private addrToLogs = new Map<Address, SigningKeyAddedOrRemovedLog[]>();
+  private addrToKeyData = new Map<Address, KeyData[]>();
+
+  constructor(private vc: ViemClient) {}
+
+  async init() {
+    await this.vc.pipeLogs(
+      {
+        event: signingKeyAddedEvent,
+      },
+      this.parseLogs
+    );
+
+    await this.vc.pipeLogs(
+      {
+        event: signingKeyRemovedEvent,
+      },
+      this.parseLogs
+    );
+    console.log(`[KEY-REG] watching logs`);
+  }
+
+  /** Get current keys of an address. */
+  async getKeys(addr: Address) {
+    try {
+      return await this.vc.publicClient.readContract({
+        abi: accountABI,
+        address: addr,
+        functionName: "getSigningKeys",
+      });
+    } catch (e: unknown) {
+      // Old Account, just ignore.
+      console.error(
+        `[API] Failed to get keys for ${addr}, probably an old account?`,
+        e
+      );
+      return [];
+    }
+  }
+
+  /** Parses account key add/remove logs, first on init() and then on subscription. */
+  parseLogs = async (logs: SigningKeyAddedOrRemovedLog[]) => {
+    const currentBlockNumber = await this.vc.publicClient.getBlockNumber(); // TODO: remove?
+
+    const addrToNewLogs: Map<Address, SigningKeyAddedOrRemovedLog[]> =
+      new Map();
+    for (const log of logs) {
+      const addr = getAddress(log.address);
+      if (addrToNewLogs.get(addr) === undefined) {
+        addrToNewLogs.set(addr, []);
+      }
+      addrToNewLogs.get(addr)!.push(log);
+    }
+
+    for (const addr of addrToNewLogs.keys()) {
+      const newLogs = addrToNewLogs.get(addr)!;
+      if (this.addrToLogs.get(addr) === undefined) {
+        this.addrToLogs.set(addr, []);
+      }
+      this.addrToLogs.get(addr)!.push(...newLogs);
+
+      const onChainKeys = (await this.getKeys(addr)).map(
+        contractFriendlyKeyToDER
+      );
+
+      this.cacheAddressProperties(addr, onChainKeys, currentBlockNumber);
+    }
+  };
+
+  /** Cache an address's key properties in memory. */
+  cacheAddressProperties = (
+    addr: Address,
+    currentKeySet: Hex[],
+    currentBlockNumber: bigint
+  ) => {
+    // deterministically sort all logs
+    const sortedLogs = this.addrToLogs.get(addr)!.sort((a, b) => {
+      const aBlockNumber = a.blockNumber || currentBlockNumber + 1n;
+      const bBlockNumber = b.blockNumber || currentBlockNumber + 1n;
```

Comment:
when does a log not have `blockNumber`?

Reply:
i think we can `assert(a.blockNumber != null && a.logIndex != null)`

Reply:
in which case this reduces to:

```ts
let diff = a.blockNumber - b.blockNumber;
if (diff !== 0) return diff;
return a.logIndex - b.logIndex;
```

Reply:
done, iirc at the time viem had a weird types bug (that's the TODO next to `currentBlockNumber` as well), that seems fixed now so